javaslang/src/main/java/javaslang/collection/HashSet.java

```diff
@@ -593,7 +596,13 @@ public T head() {
         if (isEmpty() || elements.isEmpty()) {
             return empty();
         } else {
-            return retainAll(elements);
+            int size = size();
```

Comment:
this reduces the complexity of `intersect` to the size of the smaller set (used for `containsAll` also)